### PR TITLE
[mlir][sparse] implementating stageSparseOpPass as an interface

### PR DESCRIPTION
**Pull request by PeimingLiu**

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 1]
```diff
@@ -1248,6 +1248,23 @@ LogicalResult UnaryOp::verify() {
...
 
+bool ConcatenateOp::needExtraSort() {
+  SparseTensorType dstStt = getSparseTensorType(*this);
+  if (dstStt.isAllDense() || !dstStt.isAllOrdered())
+    return false;
+
+  bool allSameOrdered = llvm::all_of(getInputs(), [dstStt](Value op) {
+    return getSparseTensorType(op).hasSameDimToLvl(dstStt);
+  });
+  // TODO: When conDim != 0, as long as conDim corresponding to  the first level
```
Comment (yinying-lisa-li) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:1259]: nit: extra space between "to" and "the"

**Review by yinying-lisa-li — APPROVED** [review]

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 2]
```diff
@@ -357,6 +359,12 @@ def SparseTensor_ConcatenateOp : SparseTensor_Op<"concatenate", [Pure]>,
      ```
    }];
 
+  let extraClassDeclaration = [{
+     // Whether the concatenate can be done by a single step or it would require
+     // an extra sort. Inherited from StageWithSortSparseOpInterface.
+     bool needExtraSort();
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:365]: same

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorInterfaces.td [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          convert csr -> unordered coo -> sort by column -> csc; and
+          concatenate csr, csc -> csr can be staged into
+          concatenate csr, csr -> unordered coo -> sort by row -> csr.
+  }];
+  let cppNamespace = "::mlir::sparse_tensor";
+  let methods = [
+    InterfaceMethod<
+    /*desc=*/"Return true if the operation needs an extra sort to produce the final result.",
+    /*retTy=*/"bool",
+    /*methodName=*/"needExtraSort",
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorInterfaces.td:30]: needsExtraSort() ?

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 4]
```diff
@@ -1248,6 +1248,23 @@ LogicalResult UnaryOp::verify() {
   return success();
 }
 
+bool ConcatenateOp::needExtraSort() {
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:1251]: and here

File: utils/bazel/llvm-project-overlay/mlir/BUILD.bazel [hunk 5]
```diff
@@ -1231,6 +1231,15 @@ td_library(
     deps = [":OpBaseTdFiles"],
 )
 
+td_library(
```
Comment (aartbik) [inline utils/bazel/llvm-project-overlay/mlir/BUILD.bazel:1234]: why is this so far away from Sparse?

I think I would prefer to just put this into the "SparseTensorTdFiles" rule, and depend on that (even though that pulls in more TD files, but that is what is generally done)

**Review by aartbik — APPROVED** [review]
Good to go once addressed

**Pull request merged**